File: LeetCode/417.Pacific_Atlantic_Water_Flow/417.Pacific_Atlantic_Water_Flow.java
class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        int m = heights.length, n = heights[0].length;
        // check if reach pacific
        boolean[][] pac = new boolean[m][n];
        // check if reach atlantic
        boolean[][] alt = new boolean[m][n];

        // iterate - first/last row
        for (int i = 0; i < n; i++) {
            bfs(0, i, heights, pac);
            bfs(m - 1, i, heights, alt);
        }
        // iterate - first/last column
        for (int i = 0; i < m; i++) {
            bfs(i, 0, heights, pac);
            bfs(i, n - 1, heights, alt);
        }

        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pac[i][j] && alt[i][j]) {
                    result.add(Arrays.asList(i, j));
                }
            }
        }

        return result;
    }

    final int[][] dirs = new int[][] { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 } };

    private void bfs(int i, int j, int[][] hs, boolean[][] reach) {
        int m = hs.length, n = hs[0].length;
        Queue<int[]> queue = new LinkedList<>();
        reach[i][j] = true;
        queue.offer(new int[] { i, j });

        while (!queue.isEmpty()) {
            int[] p = queue.poll();

            for (int[] dir : dirs) {
                int x = p[0] + dir[0], y = p[1] + dir[1];
                if (x >= 0 && x < m && y >= 0 && y < n && hs[x][y] >= hs[p[0]][p[1]] && !reach[x][y]) {
                    reach[x][y] = true;
                    queue.offer(new int[] { x, y });
                }
            }
        }
    }
}

File: LeetCode/417.Pacific_Atlantic_Water_Flow/417.Pacific_Atlantic_Water_Flow.swift
class Solution {
    func pacificAtlantic(_ heights: [[Int]]) -> [[Int]] {
        // result[i] can flow to both P & A
        // 0:unvisited, 1: P, 2: A, 3:Both: 4:None

        let m = heights.count, n = heights[0].count

        // can visited
        var pac = Array(repeating: Array(repeating: false, count: n), count: m)
        var alt = Array(repeating: Array(repeating: false, count: n), count: m)

        let dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        func bfs(_ src: (Int, Int), _ seen: inout [[Bool]]) {
            if seen[src.0][src.1] { return }

            var q = [src]
            var head = 0

            while head < q.count {
                let (x, y) = q[head]
                head += 1

                seen[x][y] = true

                for (dx, dy) in dirs {
                    let nx = dx + x, ny = dy + y
                    if nx >= 0, nx < m, ny >= 0, ny < n, !seen[nx][ny], heights[nx][ny] >= heights[x][y] {
                        seen[nx][ny] = true
                        q.append((nx, ny))
                    }
                }
            }
        }

        // Pacific, 1st row & col
        for i in 0 ..< n {
            bfs((0, i), &pac)
        }
        for j in 1 ..< m {
            bfs((j, 0), &pac)
        }

        // Atlantic, last row & col
        for i in 0 ..< n {
            bfs((m - 1, i), &alt)
        }
        for j in 0 ..< m {
            bfs((j, n - 1), &alt)
        }

        var res: [[Int]] = []
        for i in 0 ..< m {
            for j in 0 ..< n {
                if pac[i][j], alt[i][j] { res.append([i, j]) }
            }
        }

        return res
    }
}
